pkg/kstatus/observe/observers/common.go

```diff
@@ -0,0 +1,140 @@
+package observers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/observe/common"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/observe/reader"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/status"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/wait"
+)
+
+// BaseObserver provides some basic functionality needed by the observers.
+type BaseObserver struct {
+	Reader reader.ObserverReader
+
+	Mapper meta.RESTMapper
+
+	computeStatusFunc ComputeStatusFunc
+}
+
+// SetComputeStatusFunc allows for setting the function used by the observer for computing status. The default
+// value here is to use the status package. This is provided for testing purposes.
+func (b *BaseObserver) SetComputeStatusFunc(statusFunc ComputeStatusFunc) {
+	b.computeStatusFunc = statusFunc
+}
+
+// LookupResource looks up a resource with the given identifier. It will use the rest mapper to resolve
+// the version of the GroupKind given in the identifier.
+// If the resource is found, it will be returned and the observedResource will be nil. If there is an error
+// or the resource is not found, the observedResource will be returned containing information about the resource
+// and the problem.
+func (b *BaseObserver) LookupResource(ctx context.Context, identifier wait.ResourceIdentifier) (*unstructured.Unstructured, *common.ObservedResource) {
+	GVK, err := b.GVK(identifier.GroupKind)
+	if err != nil {
+		return nil, &common.ObservedResource{
+			Identifier: identifier,
+			Status:     status.UnknownStatus,
+			Error:      err,
+		}
+	}
+
+	var u unstructured.Unstructured
+	u.SetGroupVersionKind(GVK)
+	key := common.KeyForNamespacedResource(identifier)
+	err = b.Reader.Get(ctx, key, &u)
+	if err != nil && errors.IsNotFound(err) {
+		return nil, &common.ObservedResource{
+			Identifier: identifier,
+			Status:     status.NotFoundStatus,
+			Message:    "Resource not found",
+		}
+	}
+	if err != nil {
+		return nil, &common.ObservedResource{
+			Identifier: identifier,
+			Status:     status.UnknownStatus,
+			Error:      err,
+		}
+	}
+	u.SetNamespace(identifier.Namespace)
+	return &u, nil
+}
+
+// ObservedGeneratedResources provides a way to fetch the statuses for all resources of a given GroupKind
+// that match the selector in the provided resource. Typically, this is used to fetch the status of generated
+// resources.
+func (b *BaseObserver) ObserveGeneratedResources(ctx context.Context, observer ResourceObserver, object *unstructured.Unstructured,
+	gk schema.GroupKind, selectorPath ...string) (common.ObservedResources, error) {
```

Comment:
Could the observer expose which kind it is handling? It looks like technically you could pass this method e.g. a replicaSetObserver with kind="configmap" or whatever, which wouldn't ever make sense afaict. Even the selector path is basically an intrinsic property of the kind, though storing _that_ on the observer definitely feels wrong. 🤔

Reply:
I think letting the function fetch the GroupKind from the observer makes sense. I'll try that out.
The selector path is a property of the `Deployment` kind here, so I think it makes sense to have this information in the Deployment observer. When we had add support for additional resources like Job (which has an optional selector) and CronJob (which have no selector), we will need to see if there is a better way to handle looking up resources of different kinds. Same for looking up all pods that are targeted by a Service.

Reply:
So I think I punt on this for now. I tried to address this by adding a `SupportedGroupKind` function to the `ResourceObserver` interface, but that assumes that each observer only handles one GroupKind. That is not a good assumption as we will need some kind of catch-all observer for CRDs. This observer will probably implement the status primarily based on the conditions Erik has proposed. I agree we should try to make this better, but I need to look more into it.